accepted/component-model-async.md

```diff
@@ -0,0 +1,587 @@
+  `future`s, but those representation reflect an older, now-obsolete proposal,
+  so we'll need to make some updates.
+- `wasmprinter` and `wat` support for WAT<->binary conversion
+- `wasm-smith`/`wit-smith` support for test case generation (although this might
+  not be covered in the initial PR)
+
+A couple items to note for `wit-component` and `wit-parser`:
+
+- As of this writing,
+  [BuildTargets.md](https://github.com/WebAssembly/component-model/pull/378)
+  does not yet specify name mangling conventions for async features.  We'll need
+  to add new functions to the `NameMangling` trait defined in `validation.rs` to
+  support parsing the new async function names, but since there's no standard
+  for them yet, we'll provide only stub implementations for the `Standard`
+  implementation, meaning users will need to use the `Legacy` implementation
+  until a standard is defined.
+- From a subtyping perspective the `stream<T>` and `future<T>` types are
+  invariant with respect to `T`.  However, since they are anonymous types which
+  can appear anywhere in a WIT file, we need some way for bindings generators to
+  refer unambiguously to each such type when generating calls to
+  e.g. `stream.new`, `stream.drop`, etc.  I propose to handle this by
+  identifying such types using a name mangling that includes the name of a WIT
+  function in which the type appears (either as part of a parameter type or as
+  part of the return type) along with an ordinal index indicating where the type
+  appears when walking the types recursively left-to-right in depth-first order.
+    - For example, given the WIT function `foo: func(x: future<future<u32>>, y:
+      u32) -> stream<u8>`, the import `[future-new-0]foo` would indicate a call
+      to `future.new` for the type `future<u32>`.  Likewise, `[future-new-1]foo`
+      would indicate a call to `future.new` for `future<future<u32>>`, and
+      `[stream-new-2]` would indicate a call to `stream.new` for `stream<u8>`.
+
+## `wit-bindgen`
+
+As a first step, we'll add support for these new features to the Rust binding
+generator (macro and CLI) only, with `todo!()` stubs for other languages.  Based
+on user interest and resources, we can incrementally add support for other
+languages over time.
+
+For the Rust generator, we'll add an `async` option to the `generate!` macro
+(and a corresponding CLI option) to specify which imports and/or exports should
+be generated as `async` functions, e.g.:
+
+```rust
+#[derive(Default, Debug, Clone)]
+pub enum AsyncConfig {
+    /// No async imports or exports (i.e. all synchronous)
+    #[default]
+    None,
+    /// Only the specified imports and/or exports will be async
+    Some {
+        /// List of fully-qualified import names (e.g. `wasi:http/types@0.3.0-draft#[static]body.finish`)
+        imports: HashSet<String>,
+        /// List of fully-qualified export names (e.g. `wasi:http/handler@0.3.0-draft#handle`)
+        exports: HashSet<String>,
+    },
```

Comment:
How (un)likely are scenarios in which all imports, but only some exports should be async, or vice-versa? If they might be reasonably likely, it might make sense to support `All` or `None` for one, and `Some` for the other. I assume you have a much better intuition on this than I do, though